cmd/registry/cmd/label.go

```diff
@@ -0,0 +1,258 @@
+			for _, labeling := range args[1:] {
+				pair := strings.Split(labeling, "=")
+				if len(pair) != 2 {
+					log.Fatalf("%q must have the form \"key=value\" (value can be empty)", labeling)
+				}
+				if pair[0] == "" {
+					log.Fatalf("%q is invalid because it specifies an empty label", labeling)
+				}
+				labelPairs[pair[0]] = pair[1]
+			}
+		}
+
+		err = matchAndHandleLabelCmd(ctx, client, taskQueue, args[0])
+		if err != nil {
+			log.Fatalf("%s", err.Error())
+		}
+
+		close(taskQueue)
+		core.WaitGroup().Wait()
+	},
+}
+
+type labelApiTask struct {
+	ctx    context.Context
+	client connection.Client
+	api    *rpc.Api
+}
+
+type labelVersionTask struct {
+	ctx     context.Context
+	client  connection.Client
+	version *rpc.ApiVersion
+}
+
+type labelSpecTask struct {
+	ctx    context.Context
+	client connection.Client
+	spec   *rpc.ApiSpec
+}
+
+func (task *labelApiTask) Name() string {
+	return "label " + task.api.Name
+}
+
+func (task *labelApiTask) Run() error {
+	if task.api.Labels == nil {
+		task.api.Labels = make(map[string]string)
+	}
+	if len(labelPairs) > 0 {
+		log.Printf("label %s with %v", task.api.Name, labelPairs)
+		for k, v := range labelPairs {
+			if v == "" {
+				delete(task.api.Labels, k)
+			} else {
+				task.api.Labels[k] = v
+			}
+		}
+		req := &rpc.UpdateApiRequest{
+			Api: task.api,
+			UpdateMask: &field_mask.FieldMask{
+				Paths: []string{"labels"},
+			},
+		}
+		_, err := task.client.UpdateApi(task.ctx, req)
+		return err
+	}
+	return nil
+}
+
+func (task *labelVersionTask) Name() string {
+	return "label " + task.version.Name
+}
+
+func (task *labelVersionTask) Run() error {
+	if task.version.Labels == nil {
+		task.version.Labels = make(map[string]string)
+	}
+	if len(labelPairs) > 0 {
+		log.Printf("label %s with %v", task.version.Name, labelPairs)
+		for k, v := range labelPairs {
+			if v == "" {
+				delete(task.version.Labels, k)
+			} else {
+				task.version.Labels[k] = v
+			}
+		}
+		req := &rpc.UpdateApiVersionRequest{
+			ApiVersion: task.version,
+			UpdateMask: &field_mask.FieldMask{
+				Paths: []string{"labels"},
+			},
+		}
+		_, err := task.client.UpdateApiVersion(task.ctx, req)
+		return err
+	}
+	return nil
+}
+
+func (task *labelSpecTask) Name() string {
+	return "label " + task.spec.Name
+}
+
+func (task *labelSpecTask) Run() error {
+	if task.spec.Labels == nil {
+		task.spec.Labels = make(map[string]string)
+	}
+	if len(labelPairs) > 0 {
+		log.Printf("label %s with %v", task.spec.Name, labelPairs)
+		for k, v := range labelPairs {
+			if v == "" {
+				delete(task.spec.Labels, k)
+			} else {
+				task.spec.Labels[k] = v
+			}
+		}
+		req := &rpc.UpdateApiSpecRequest{
+			ApiSpec: task.spec,
+			UpdateMask: &field_mask.FieldMask{
+				Paths: []string{"labels"},
+			},
+		}
+		_, err := task.client.UpdateApiSpec(task.ctx, req)
+		return err
+	}
+	return nil
+}
+
+func matchAndHandleLabelCmd(
+	ctx context.Context,
+	client connection.Client,
+	taskQueue chan core.Task,
```

Comment:
Nit: Since we only want this function to add tasks to the queue (not read them) we can indicate that using directions, e.g.`taskQueue chan<- core.Task,`. Same for the functions below.

Reply:
Thanks, I'll change these. This usage exists throughout the registry tool, so it will be inconsistent without an overall fix.